rholang/src/test/scala/coop/rchain/rholang/interpreter/ReplaySpec.scala

```diff
@@ -0,0 +1,131 @@
+package coop.rchain.rholang.interpreter
+
+import cats.effect.Resource
+import cats.syntax.all._
+import coop.rchain.crypto.hash.Blake2b512Random
+import coop.rchain.metrics
+import coop.rchain.metrics.{Metrics, NoopSpan, Span}
+import coop.rchain.rholang.Resources
+import coop.rchain.rholang.interpreter.accounting.utils.costLog
+import coop.rchain.rholang.interpreter.accounting.{_cost, Cost, CostAccounting}
+import coop.rchain.rspace.SoftCheckpoint
+import coop.rchain.shared.Log
+import monix.eval.Task
+import monix.execution.Scheduler.Implicits.global
+import org.scalatest.{FlatSpec, Matchers}
+
+import scala.concurrent.duration._
+
+class ReplaySpec extends FlatSpec with Matchers {
+
+  "multi joins (2/2)" should "execute successfully in replay" in {
+    val term =
+      """
+        |new x in {
+        |  x!() | for(<- x; <- x) { 0 } | x!()
+        |}
+        |""".stripMargin
+    testRholangTerm(term, 500, 30.seconds)
+  }
+
+  "multi joins (4/2)" should "execute successfully in replay" in {
+    val term =
+      """
+        |new x in {
+        |  x!() | x!() | for(<- x; <- x) { 0 } | x!() | x!()
+        |}
+        |""".stripMargin
+    testRholangTerm(term, 500, 30.seconds)
+  }
+
+  def testRholangTerm(term: String, iterations: Int, timeout: Duration) =
```

Comment:
Very nice!

Reply:
One unfortunate thing, although it's completely executed in-memory, it cannot be run in parallel. First because of _Cost_ instance and also because of _replayData_ in `IReplaySpace`.

But if we could make snapshot of the whole `Runtime`, we could make a fork and execute terms in parallel from the same parent point.
https://github.com/rchain/rchain/blob/f7e46a9346990010bf32422f0c1abee17a1f85c5/rholang/src/main/scala/coop/rchain/rholang/interpreter/Runtime.scala#L31-L39